### ISSUE TITLE
OpenAPI on calculated field add an extra `items` fields even if `openApiContext` define an object

### ISSUE THREAD

**Issue by monitaurus** [api-platform/core#6391]
**API Platform version(s) affected**: 3.2.22, 3.3.2

Tested on 3.2.22 and replicated on 3.3.2

**Description**  
On a calculated field that returns an `array`, if we add a `ApiProperty` to change the type into `object`, the generated OpenAPI does generate a `"type": "object"` but there is also an extra `"items": {"type": "string"}`.

**How to reproduce**  

On a resource, add the following calculated field:
```php
    #[ApiProperty(
        openapiContext: [
            'type' => 'object',
            'properties' => [
                'alpha' => ['type' => 'integer'],
                'beta' => ['type' => 'string'],
            ],
        ]
    )]
    public function getCalculatedField(): array
    {
        return [
            'alpha' => 123,
            'beta' => 'hello',
        ];
    }
```

Then access OpenAPI via Swagger, or generate a openapi json to find the extra `items`.

![swagger_screenshot](https://github.com/api-platform/core/assets/1794571/3f6fa67c-e179-4f2d-b4eb-96c0ba22110d)

**Possible Solution**  

None found at the moment.

**Additional Context**  

Investigation on the factory code:
* the process revolve around `SchemaPropertyMetadataFactory::create`
* the calculated field returns an array, so `$propertyMetadata->getBuiltinTypes()` returns `["array"]` ([view file](https://github.com/api-platform/core/blob/e867d07f59b82d5f1bdca69e096ddf452dd7efc8/src/JsonSchema/Metadata/Property/Factory/SchemaPropertyMetadataFactory.php#L86))
* `array` is considered as a collection of items, and the code define items as `string` by default ([view file](https://github.com/api-platform/core/blob/e867d07f59b82d5f1bdca69e096ddf452dd7efc8/src/JsonSchema/Metadata/Property/Factory/SchemaPropertyMetadataFactory.php#L121))
* and finally `$this->getType` returns a property type as `array` with `items` as `string` ([view file](https://github.com/api-platform/core/blob/e867d07f59b82d5f1bdca69e096ddf452dd7efc8/src/JsonSchema/Metadata/Property/Factory/SchemaPropertyMetadataFactory.php#L133))

The type `object` seems to override the type `array` partially.
And as builtin types are computed from PHP type ([see doc](https://api-platform.com/docs/v3.0/core/upgrade-guide/#:~:text=Note%20that%20builtinTypes%20are%20computed%20automatically%20from%20PHP%20types.)), I don't think editing `ApiProperty` is enough to fix the issue.

**Comment by monitaurus** [comment]
On a personal note, If it's really issue, I can take time to fix it, with some insights or guidance.

**Comment by soyuka** [comment]
it should return at

**Comment by soyuka** [comment]
IMO as you customized the schema it should return early at: 

https://github.com/api-platform/core/blob/e867d07f59b82d5f1bdca69e096ddf452dd7efc8/src/JsonSchema/Metadata/Property/Factory/SchemaPropertyMetadataFactory.php#L107

Definitely a bug, would love for this to get fixed, lmk if I can assist you (target 3.2 branch)

**Comment by monitaurus** [comment]
I'll take a look next week, many thanks for the details 🙇
